azkaban-web-server/src/main/java/cloudflow/services/FlowServiceImpl.java

```diff
@@ -0,0 +1,200 @@
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+import azkaban.user.User;
+import cloudflow.daos.FlowDao;
+import cloudflow.daos.FlowVersionDao;
+import cloudflow.daos.ProjectDao;
+import cloudflow.error.CloudFlowException;
+import cloudflow.error.CloudFlowNotFoundException;
+import cloudflow.error.CloudFlowValidationException;
+import cloudflow.models.FlowResponse;
+import cloudflow.models.FlowVersion;
+import cloudflow.models.Project;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class FlowServiceImpl implements FlowService {
+
+  private static final String PROJECT_ID_PARAM = "project_id";
+  private static final String PROJECT_VERSION_PARAM = "project_version";
+  private static final String PROJECT_NAME_PARAM = "project_name";
+  private static final String EXPAND_PARAM = "expand";
+  private static final String VERSIONS_KEY = "versions";
+  private static final Logger log = LoggerFactory.getLogger(ProjectServiceImpl.class);
+  private final FlowDao flowDao;
+  private final FlowVersionDao flowVersionDao;
+  private final ProjectDao projectDao;
+
+  @Inject
+  public FlowServiceImpl(FlowDao flowDao, FlowVersionDao flowVersionDao, ProjectDao projectDao) {
+    this.flowDao = flowDao;
+    this.projectDao = projectDao;
+    this.flowVersionDao = flowVersionDao;
+  }
+
+
+  @Override
+  public FlowResponse getFlow(String flowId, Map<String, String[]> queryParamMap) {
+    requireNonNull(flowId, "Flow id cannot be null");
+    requireNonNull(queryParamMap, "query param map cannot be null");
+
+    Optional<String> expandValue = Optional.empty();
+
+    for (Map.Entry<String, String[]> entry : queryParamMap.entrySet()) {
+      String key = entry.getKey();
+      String[] value = entry.getValue();
+      if (key.equals(EXPAND_PARAM)) {
+        expandValue = Optional.of(extractArgumentValue(key, value));
+
+        /*Currently only expand=versions is supported */
+        if (!expandValue.get().equals(VERSIONS_KEY)) {
+          throw new CloudFlowValidationException("Invalid value for expand: " + expandValue);
+        }
+      }
+    }
+
+    Optional<FlowResponse> flow = flowDao.getFlow(flowId);
+    if (!flow.isPresent()) {
+      String errorMsg = "Flow does not exist for id: " + flowId;
+      log.error(errorMsg);
+      throw new CloudFlowNotFoundException(errorMsg);
+    }
+
+    setFlowCreateAndLastModifiedInfo(flow.get());
+    setLastFlowVersion(flow.get());
+
+    /* If expand=versions is present in query params, get all the versions for the flow */
+    if (expandValue.isPresent()) {
+      flow.get().setFlowVersions(new ArrayList<>());
+      flow.get().setFlowVersions(flowVersionDao.getAllVersions(flowId));
+    }
+
+    return flow.get();
+  }
+
+
+  private String extractArgumentValue(String argName, String[] argValues) {
```

Comment:
Other endpoints have similar methods for validation. Should consolidate and extract it to a common utility class. Let me know what you think about it.

Reply:
I was thinking of the same thing while using it. Didn't want to combine the refactoring with this change and we need to discussion Utility class contents, structure, etc. Will create a ticket to do this as a follow-up.

Reply:
Created the ticket: https://jira01.corp.linkedin.com:8443/browse/LIHADOOP-51813